nvtabular/io.py

```diff
@@ -412,39 +504,87 @@ def add_data(self, gdf):
+            fn = path.split(self.fs.sep)[-1]
+            data["file_stats"].append({"file_name": fn, "num_rows": self.num_samples[i]})
+        # cats
+        data["cats"] = []
+        for c in self.cats:
+            data["cats"].append({"col_name": c, "index": self.col_idx[c]})
+        # conts
+        data["conts"] = []
+        for c in self.conts:
+            data["conts"].append({"col_name": c, "index": self.col_idx[c]})
+        # labels
+        data["labels"] = []
+        for c in self.labels:
+            data["labels"].append({"col_name": c, "index": self.col_idx[c]})
 
-    def _write_filelist(self):
-        file_list_writer = open(os.path.join(self.out_dir, "file_list.txt"), "w")
-        file_list_writer.write(str(self.num_out_files) + "\n")
-        for f in self.data_files:
+        return data
+
+    @classmethod
+    def write_general_metadata(cls, data, fs, out_dir):
+        if not data:
+            return
+        data_paths = data.pop("data_paths", [])
+        num_out_files = len(data_paths)
+
+        # Write file_list
+        file_list_writer = fs.open(fs.sep.join([out_dir, "file_list.txt"]), "w")
+        file_list_writer.write(str(num_out_files) + "\n")
+        for f in data_paths:
             file_list_writer.write(f + "\n")
         file_list_writer.close()
 
-    def close(self):
+        # Write metadata json
+        metadata_writer = fs.open(fs.sep.join([out_dir, "metadata.json"]), "w")
+        json.dump(data, metadata_writer)
+        metadata_writer.close()
+
+    @classmethod
+    def write_special_metadata(cls, data, fs, out_dir):
+        pass
+
+    def _close_writers(self):
+        for writer in self.data_writers:
+            writer.close()
+        return None
+
+    def close(self, write_metadata=True):
         # wake up all the worker threads and signal for them to exit
         for _ in range(self.num_threads):
             self.queue.put(self._eod)
 
         # wait for pending writes to finish
         self.queue.join()
 
-        self._write_filelist()
-        self._write_metadata()
-
-        # Close writers
-        for writer in self.data_writers:
-            writer.close()
+        # Close writers and collect various metadata
+        _gen_meta = self.package_general_metadata()
+        _special_meta = self._close_writers()
+        return _gen_meta, _special_meta
 
 
 class ParquetWriter(ThreadedWriter):
-    def __init__(
-        self, out_dir, num_out_files=30, num_threads=4, cats=None, conts=None, labels=None
-    ):
-        super().__init__(out_dir, num_out_files, num_threads, cats, conts, labels)
-        self.data_files = [os.path.join(out_dir, f"{i}.parquet") for i in range(num_out_files)]
-        self.data_writers = [pwriter(f, compression=None) for f in self.data_files]
+    def __init__(self, out_dir, use_guid=False, **kwargs):
+        super().__init__(out_dir, **kwargs)
+        self.data_fns = []
+        self.data_paths = []
```

Comment:
Why do we need to store both data_fns and data_paths?

Reply:
We don't need to, but it was convenient to access the file names this way (when updating the metadata for the global _metadata file) -- I'll see if I can just clip the file name off the end of the full path.

Reply:
I removed `self.data_fns`